src/v/cluster/distributed_kv_stm.h

```diff
@@ -0,0 +1,422 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/fwd.h"
+#include "cluster/logger.h"
+#include "cluster/persisted_stm.h"
+#include "cluster/simple_batch_builder.h"
+#include "serde/serde.h"
+
+namespace cluster {
+
+/**
+ * Implements simple distributed KV store like semantics using the state
+ * machine abstraction. Backed by a Kafka topic with one or more partitions.
+ *
+ * Supports the following operations
+ *
+ * add_kv(key, val)
+ * delete_kv(key)
+ * value_for(key)
+ * coordinator_for(key) - only on routing partition (discussed below)
+ *
+ * Input KV pairs are spread across the partitions of the topic for
+ * load balancing, this requires that each key K to be tied to a partition.
+ * Such a partition is called coordinator. A coordinator for a key K is the
+ * id of the partition responsible for storage and querying KV mapping for K.
+
+ * An easy way to map a key -> coordinator is using a static hash function but
+ * that doesn't give us the flexibility to increase the partition to scale
+ * the operations. For example, we start with 1 partition, realize it is a
+ * bottleneck and want to increase to 3 partitions, the hash function is
+ * stale. To have this flexibility of scaling partitions, we use a routing
+ * partition that tracks the coordinator mappings. It is partition 0 of the
+ * topic that also works as a coordinator along with routing capabilities.
+ *
+ * Routing partition keeps track of repartitioning operations and remembers
+ * the coordinator mappings prior to repartitiong. So any new keys are hashed
+ * across a larger set of partitions while still retaining original mappings.
+ * Migration of keys across coordinators is not supported and can be added
+ * later on.
+ *
+ * The query pattern from clients is expected to be as follows.
+ * - query routing_partition using coordinator_for(key) to get the coordinator
+ *   for key.
+ * - call add_kv(key, val)/delete_kv(key)/value_for(key) on the coordinator.
+ *
+ * For a reference client implementation, look at transforms_offsets topic
+ * that uses this KV store to track consumption offsets for transforms.
+ */
+
+template<class T>
+concept SerdeSerializable = requires(T t, iobuf buf, iobuf_parser parser) {
+    serde::write_async(buf, t);
+    serde::read_async<T>(parser);
+    serde::from_iobuf<T>(std::move(buf));
+    serde::to_iobuf(t);
+};
+
+template<SerdeSerializable Key, SerdeSerializable Value>
+class kv_stm final : public persisted_stm<> {
+public:
+    explicit kv_stm(
+      size_t max_partitions, ss::logger& logger, raft::consensus* raft)
+      : persisted_stm<>("kv_stm.snapshot", logger, raft)
```

Comment:
since this is generic we should parametrize state machine name and snapshot name to make it possible to create multiple instances of this stm

Reply:
hmm, maybe you are right and we do not need it.